### PR TITLE
uvicorn appid improvement

### PR DESCRIPTION
**Pull request by pfilipko1**
https://github.com/Granulate/gprofiler/issues/693

### CODE REVIEW

File: gprofiler/metadata/application_identifiers.py [hunk 1]
```diff
@@ -145,6 +145,53 @@ def get_app_id(self, process: Process) -> Optional[str]:
         return None
 
 
+class _UvicornApplicationIdentifierBase(_ApplicationIdentifier):
+    def uvicorn_to_app_id(self, wsgi_app_spec: str, process: Process) -> str:
+        wsgi_app_file = wsgi_app_spec.split(":", maxsplit=1)[0]
+        return f"uvicorn: {wsgi_app_spec} ({_append_python_module_to_proc_wd(process, wsgi_app_file)})"
+
+    def uvicorn_get_app_name(self, cmdline_list: List[str]) -> str:
```
Comment (Jongy) [inline gprofiler/metadata/application_identifiers.py:153]: [Same here](https://github.com/Granulate/gprofiler/pull/816#discussion_r1246554310)

File: gprofiler/metadata/application_identifiers.py [hunk 2]
```diff
@@ -145,6 +145,53 @@ def get_app_id(self, process: Process) -> Optional[str]:
...
+        if "uvicorn" != os.path.basename(_get_cli_arg_by_index(process.cmdline(), 0)) and "uvicorn" != os.path.basename(
+            _get_cli_arg_by_index(process.cmdline(), 1)
+        ):
+            return None
+
+        # app specification does not have to be last, thus a method to extract it from commandline
+        return self.uvicorn_to_app_id(self.uvicorn_get_app_name(process.cmdline()), process)
+
+
+class _UvicornTitleApplicationIdentifier(_UvicornApplicationIdentifierBase):
```
Comment (Jongy) [inline gprofiler/metadata/application_identifiers.py:176]: Is it relevant to uvicorn as well? I mean, usage of setproctitle, like gunicorn does.

File: gprofiler/metadata/application_identifiers.py [hunk 3]
```diff
@@ -145,6 +145,53 @@ def get_app_id(self, process: Process) -> Optional[str]:
         return None
 
 
+class _UvicornApplicationIdentifierBase(_ApplicationIdentifier):
+    def uvicorn_to_app_id(self, wsgi_app_spec: str, process: Process) -> str:
+        wsgi_app_file = wsgi_app_spec.split(":", maxsplit=1)[0]
+        return f"uvicorn: {wsgi_app_spec} ({_append_python_module_to_proc_wd(process, wsgi_app_file)})"
+
+    def uvicorn_get_app_name(self, cmdline_list: List[str]) -> str:
+        # factory argument is being checked explicitly, because it contains minuses and precedes appid
```
Comment (Jongy) [inline gprofiler/metadata/application_identifiers.py:154]: Btw where did you see this `--factory` occurring?

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: gprofiler/metadata/application_identifiers.py (hunk 3 above)
Comment (pfilipko1) [reply gprofiler/metadata/application_identifiers.py:154]: https://www.uvicorn.org/#application-factories

File: gprofiler/metadata/application_identifiers.py (hunk 2 above)
Comment (pfilipko1) [reply gprofiler/metadata/application_identifiers.py:176]: Need to check, will check :)

Comment (pfilipko1) [reply gprofiler/metadata/application_identifiers.py]: Checked, no such method, removed the title identifier.

File: gprofiler/metadata/application_identifiers.py [hunk 4]
```diff
@@ -145,6 +145,38 @@ def get_app_id(self, process: Process) -> Optional[str]:
...
+class _UvicornApplicationIdentifierBase(_ApplicationIdentifier):
+    def uvicorn_to_app_id(self, wsgi_app_spec: str, process: Process) -> str:
+        wsgi_app_file = wsgi_app_spec.split(":", maxsplit=1)[0]
+        return f"uvicorn: {wsgi_app_spec} ({_append_python_module_to_proc_wd(process, wsgi_app_file)})"
+
+    def uvicorn_get_app_name(self, cmdline_list: List[str]) -> str:
+        # method improving appid selection:
+        # https://github.com/Granulate/gprofiler/issues/693
+        # factory argument is being checked explicitly, because it contains minuses and may precede appid
+        cmdline_colon_list = list(filter(lambda x: ":" in x, cmdline_list))
```
Comment (Jongy) [inline gprofiler/metadata/application_identifiers.py:157]: Same comment on cmdline_colon_list as I gave [here](https://github.com/Granulate/gprofiler/pull/816)

File: tests/test_appids.py
Comment (Jongy) [inline tests/test_appids.py:1]: As suggested in https://github.com/Granulate/gprofiler/pull/816 let's add a test case for the "unknown app name" here as well.

**Review by Jongy — CHANGES_REQUESTED** [review]
.

**Review by Jongy — APPROVED** [review]

**Pull request merged**